src/DotNetWorker.Grpc/GrpcServiceCollectionExtensions.cs

```diff
@@ -56,40 +49,21 @@ public static IServiceCollection AddGrpc(this IServiceCollection services)
 
             // gRPC Core services
             services.AddSingleton<IWorker, GrpcWorker>();
-            services.AddSingleton<FunctionRpcClient>(p =>
-            {
-                IOptions<GrpcWorkerStartupOptions> argumentsOptions = p.GetService<IOptions<GrpcWorkerStartupOptions>>()
-                    ?? throw new InvalidOperationException("gRPC Services are not correctly registered.");
-
-                GrpcWorkerStartupOptions arguments = argumentsOptions.Value;
-
-                string uriString = $"http://{arguments.Host}:{arguments.Port}";
-                if (!Uri.TryCreate(uriString, UriKind.Absolute, out Uri? grpcUri))
-                {
-                    throw new InvalidOperationException($"The gRPC channel URI '{uriString}' could not be parsed.");
-                }
-
 
 #if NET5_0_OR_GREATER
-                GrpcChannel grpcChannel = GrpcChannel.ForAddress(grpcUri, new GrpcChannelOptions()
-                {
-                    MaxReceiveMessageSize = arguments.GrpcMaxMessageLength,
-                    MaxSendMessageSize = arguments.GrpcMaxMessageLength,
-                    Credentials = ChannelCredentials.Insecure
-                });
+            // If we are running in the native host process, use the native client
+            // for communication (interop). Otherwise; use the gRPC client.
+            if (AppContext.GetData("AZURE_FUNCTIONS_NATIVE_HOST") is not null)
+            {
+                services.AddSingleton<IWorkerClientFactory, Azure.Functions.Worker.Grpc.NativeHostIntegration.NativeWorkerClientFactory>();
+            }
+            else
+            {
+                services.AddSingleton<IWorkerClientFactory, GrpcWorkerClientFactory>();
+            }
 #else
-
-                var options = new ChannelOption[]
-                {
-                    new ChannelOption(Grpc.Core.ChannelOptions.MaxReceiveMessageLength, arguments.GrpcMaxMessageLength),
-                    new ChannelOption(Grpc.Core.ChannelOptions.MaxSendMessageLength, arguments.GrpcMaxMessageLength)
-                };
-
-                Grpc.Core.Channel grpcChannel = new Grpc.Core.Channel(arguments.Host, arguments.Port, ChannelCredentials.Insecure, options);
-
+            services.AddSingleton<IWorkerClientFactory, GrpcWorkerClientFactory>();
```

Comment:
Will NETFX case falls into this else case? We want the nativehost in NetFX as well. correct?

Reply:
No native host for NetFX